### PR TITLE
feat: support Llama Parse

### PR DESCRIPTION
**Pull request by leehuwuj**

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -59,4 +61,6 @@ async def redirect():
     app_port = int(os.getenv("APP_PORT", "8000"))
     reload = environment == "dev"
 
-    uvicorn.run(app="main:app", host=app_host, port=app_port, reload=reload)
+    uvicorn.run(
+        app="main:app", host=app_host, port=app_port, reload=reload, loop="asyncio"
```
Comment (leehuwuj) [inline main.py:65]: We need to use the `asyncio` loop to enable the nested feature (which is necessary for llama parse).

File: patch/backend/app/engine/loaders/file.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    parser = LlamaParse(result_type="markdown", verbose=True, language="en")
+    return parser
+
+
+def get_file_documents(config: FileLoaderConfig):
+    from llama_index.core.readers import SimpleDirectoryReader
+
+    import nest_asyncio
+
+    nest_asyncio.apply()
```
Comment (leehuwuj) [inline patch/backend/app/engine/loaders/file.py:35]: Apply nest_asyncio. See: https://github.com/run-llama/llama_parse?tab=readme-ov-file#getting-started

File: src/controllers/loader.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+from pydantic import Field, validator
+from pydantic_settings import BaseSettings
+from typing import Dict, Tuple, List
+
+from src.models.loader import LoaderConfig, FileLoader
+from src.models.env_config import get_config
+from src.constants import LOADER_CONFIG_FILE
+
+
+class LoaderManager:
```
Comment (leehuwuj) [inline src/controllers/loader.py:13]: We can just implement File loader config but introducing an abstract config to manage the loader.yaml is not a bad idea.

File: patch/backend/app/engine/loaders/file.py (hunk 2 above)
Comment (marcusschiesser) [reply patch/backend/app/engine/loaders/file.py:35]: ideally we do this without patch - if we need a patch, better release a new upstream first

File: src/models/loader.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        """
+        Update the environment variable for the API key.
+        """
+        import dotenv
+
+        if self.llama_cloud_api_key:
+            # Update runtime environment variable
+            os.environ["LLAMA_CLOUD_API_KEY"] = self.llama_cloud_api_key
+            # Update .env file
+            dotenv.set_key(
```
Comment (marcusschiesser) [inline src/models/loader.py:36]: reading this, i think it makes sense to encapsulate the API keys that are needed by loaders to the file loader model.
similar we might want to encapsulate env keys for tool configuration in the tool model later.

but one question: how does this interfere with the env config model? we have to ensure that both models (env config and file loader) can be serialized and deserialized independently without impact. e.g. if the env config is serialized we don't want `LLAMA_CLOUD_API_KEY` to be overwritten.

Comment (leehuwuj) [reply src/models/loader.py:36]: Each model just manages the environment variables linked with its fields, so they are independent.

File: patch/backend/app/engine/loaders/file.py (hunk 2 above)
Comment (leehuwuj) [reply patch/backend/app/engine/loaders/file.py]: It's now using create-llama 0.1.10 so i removed the patch.

File: admin/sections/fileLoader.tsx [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        title: "Config updated successfully",
+      });
+    } catch (error) {
+      console.error(error);
+      toast({
+        className: cn(
+          "top-0 right-0 flex fixed md:max-w-[420px] md:top-4 md:right-4 text-red-500",
+        ),
+        title: "Failed to update file loader configuration!",
+      });
```
Comment (marcusschiesser) [inline admin/sections/fileLoader.tsx:55]: @leehuwuj the toasts are always the same right?  if so how about sending a PR after the release which moves them to a utils class, e.g.:
```suggestion
      toastSuccess({
        title: "Config updated successfully",
      });
    } catch (error) {
      console.error(error);
      toastFail({
        title: "Failed to update file loader configuration!",
      });
```

**Review by marcusschiesser — APPROVED** [review]

**Pull request merged**